app/src/main/java/org/mozilla/fenix/components/metrics/AppAllSourceStartTelemetry.kt

```diff
@@ -0,0 +1,59 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.metrics
+
+import android.content.Intent
+import mozilla.components.support.utils.SafeIntent
+import mozilla.components.support.utils.toSafeIntent
+
+/**
+ * Tracks how the application was opened through [Event.AppOpenedAllSourceStartup]
+ * we only track the event if app is launching first time or was in background state because
+ * we don't want random intents with [Intent.ACTION_VIEW] to be tracked as app opened
+ * For example, Start app through APP_ICON -> run adb intent command to open a link.
+ * We want to ignore these adb commands since app is already on foreground.
+ */
+class AppAllSourceStartTelemetry(
+    intent: Intent?,
+    private val metrics: MetricController,
+    private val isExternalAppBrowserActivity: Boolean
+) {
+
+    /**
+     * Track the event on a cold startup
+     */
+    init {
+        intent?.toSafeIntent()?.let(::getIntentAllStartSource)
+            ?.also {
+                metrics.track(Event.AppOpenedAllSourceStartup(it))
```

Comment:
This code is inconsistent: why do we track one event as part of object initialization and another as part of a function call?

I think it'd be simpler to track both events in functions – code run in object initialization and passed into the initializer can impact the entire class but a function on a class with no mutable variables will generally only impact itself.

Additionally, in general I don't like creating side effects in object initialization but to be honest, I'm struggling to put my finger on why. Conceptually, I think of object initialization as a function to allocate memory and configure an object: I guess it makes the method have more than one responsibility to do more than that. It also makes the object harder to mock (if applicable) because the initializer will run in some mocking frameworks before other methods. It's also less flexible – e.g. if you block the thread to init an object, you can't use the object concurrently until it's done. Unfortunately, I guess it's just a few hand-wavy reasons for this so maybe I'm wrong but I still think we're handling these inconsistently and that handling the functionality through functions is simpler.

---

If we handle this as a function, we can further simplify this by initializing this object as part of `Components`. This is simpler because the lifecycle of the object becomes the same as the Application, i.e. for practical purposes, forever, instead of the lifecycle of the Activity.

Reply:
That makes sense. I will update this to make it more consistent so that we dont have side effects on init.

Reply:
So I moved the `LifecycleObserver` inside `AppAllSourceStartTelemetry`. The only thing inside init now is attaching our class to `ProcessLifecycleOwner` 's observers list.